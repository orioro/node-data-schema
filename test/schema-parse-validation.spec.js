import {
  schemaParseValidation,
} from '../src/schema-parse-validation'

describe('schemaParseValidation(options, schema)', () => {

  test('not recursive', () => {
    const SCHEMA = {
      label: 'Document',
      type: 'map',
      required: true,
      attributes: {
        title: {
          type: 'text',
          required: true,
        },
        description: {
          type: 'richText',
          required: true,
        },
      }
    }

    expect(schemaParseValidation({ recursive: false }, SCHEMA)).toEqual({
      notNull: {
        message: 'Document is required',
      },
      notUndefined: {
        message: 'Document is required',
      },
      objectPlain: {
        message: 'Document must be a plain object'
      }
    })
  })

  test('recursive map', () => {
    const SCHEMA = {
      type: 'map',
      attributes: {
        title: {
          type: 'text',
          required: true,
        },
        description: {
          type: 'richText',
          required: true,
        },
        place: {
          type: 'map',
          attributes: {
            name: {
              type: 'text',
              required: true,
            },
            address: {
              type: 'map',
              attributes: {
                line_1: {
                  type: 'text',
                  validation: {
                    stringMinLength: {
                      length: 10,
                    },
                    stringMaxLength: {
                      length: 500,
                    }
                  }
                },
              },
              required: true
            }
          }
        }
      },
      validation: {
        objectPlain: {
          message: 'must be a plain object'
        }
      }
    }

    const validation = schemaParseValidation({ recursive: true }, SCHEMA)

    const REQUIRED_VALIDATION = {
      notNull: {
        message: 'Value is required',
      },
      notUndefined: {
        message: 'Value is required'
      }
    }

    const PLAIN_OBJECT_VALIDATION = {
      objectPlain: {
        message: 'Value must be a plain object',
      }
    }

    expect(validation).toMatchSnapshot()
    expect(validation).toEqual({
      objectProperties: {
        properties: {
          title: {
            ...REQUIRED_VALIDATION
          },
          description: {
            ...REQUIRED_VALIDATION,
          },
          place: {
            ...PLAIN_OBJECT_VALIDATION,
            objectProperties: {
              properties: {
                name: {
                  ...REQUIRED_VALIDATION,
                },
                address: {
                  ...PLAIN_OBJECT_VALIDATION,
                  ...REQUIRED_VALIDATION,
                  objectProperties: {
                    properties: {
                      line_1: {
                        stringMinLength: {
                          length: 10,
                        },
                        stringMaxLength: {
                          length: 500,
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      },
      objectPlain: {
        message: 'must be a plain object'
      }
    })

    // console.log(JSON.stringify(validation, null, '  '))
  })

  test('recursive list', () => {
    const SCHEMA = {
      type: 'map',
      attributes: {
        list_of_addresses: {
          type: 'list',
          item: {
            type: 'map',
            attributes: {
              line_1: {
                type: 'text',
                required: true,
                validation: {
                  stringMinLength: {
                    length: 10,
                    message: 'Line 1 at least 10',
                  }
                }
              },
              line_2: {
                type: 'text',
                required: true
              },
              comments: {
                type: 'list',
                item: {
                  type: 'map',
                  attributes: {
                    author: {
                      type: 'text',
                      required: true,
                    },
                    contents: {
                      type: 'richText',
                      required: true,
                    },
                  }
                },
                validation: {
                  arrayMaxLength: {
                    length: 10,
                    message: 'At most 10 comments per address',
                  }
                }
              }
            }
          }
        }
      }
    }

    const validation = schemaParseValidation({ recursive: true }, SCHEMA)
    expect(validation).toMatchSnapshot()

    // console.log(JSON.stringify(validation, null, '  '))
  })

  test('conditionals', () => {
    const SCHEMA = {
      type: 'map',
      attributes: {
        name: {
          type: 'string',
        },
        age: {
          type: 'number',
        },
        legalResponsible: {
          type: 'string',
        }
      },
      conditionals: [
        {
          criteria: {
            age: {
              $lt: 18
            }
          },
          attributes: {
            legalResponsible: {
              required: true,
            }
          }
        }
      ]
    }

    const validation = schemaParseValidation({ recursive: true }, SCHEMA)

    // console.log(JSON.stringify(validation, null, '  '))
  })

  test('conditionals - 2', () => {
    const SCHEMA = {
      type: 'number',
      conditionals: [
        {
          criteria: {
            $gt: 10
          },
          validation: {
            numberMultipleOf: {
              base: 15,
            }
          }
        }
      ]
    }

    const validation = schemaParseValidation({ recursive: false }, SCHEMA)

    // console.log(JSON.stringify(validation, null, '  '))
  })
})
