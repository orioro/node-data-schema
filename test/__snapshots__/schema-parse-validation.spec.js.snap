// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaParseValidation(options, schema) recursive list 1`] = `
Object {
  "objectPlain": Object {
    "message": "Value must be a plain object",
  },
  "objectProperties": Object {
    "properties": Object {
      "list_of_addresses": Object {
        "array": Object {
          "message": "Value must be an array",
        },
        "arrayItems": Object {
          "validation": Object {
            "objectPlain": Object {
              "message": "Value must be a plain object",
            },
            "objectProperties": Object {
              "properties": Object {
                "comments": Object {
                  "array": Object {
                    "message": "Value must be an array",
                  },
                  "arrayItems": Object {
                    "validation": Object {
                      "objectPlain": Object {
                        "message": "Value must be a plain object",
                      },
                      "objectProperties": Object {
                        "properties": Object {
                          "author": Object {
                            "notNull": Object {
                              "message": "Value is required",
                            },
                            "notUndefined": Object {
                              "message": "Value is required",
                            },
                          },
                          "contents": Object {
                            "notNull": Object {
                              "message": "Value is required",
                            },
                            "notUndefined": Object {
                              "message": "Value is required",
                            },
                          },
                        },
                      },
                    },
                  },
                  "arrayMaxLength": Object {
                    "length": 10,
                    "message": "At most 10 comments per address",
                  },
                },
                "line_1": Object {
                  "notNull": Object {
                    "message": "Value is required",
                  },
                  "notUndefined": Object {
                    "message": "Value is required",
                  },
                  "stringMinLength": Object {
                    "length": 10,
                    "message": "Line 1 at least 10",
                  },
                },
                "line_2": Object {
                  "notNull": Object {
                    "message": "Value is required",
                  },
                  "notUndefined": Object {
                    "message": "Value is required",
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`schemaParseValidation(options, schema) recursive map 1`] = `
Object {
  "objectPlain": Object {
    "message": "must be a plain object",
  },
  "objectProperties": Object {
    "properties": Object {
      "description": Object {
        "notNull": Object {
          "message": "Value is required",
        },
        "notUndefined": Object {
          "message": "Value is required",
        },
      },
      "place": Object {
        "objectPlain": Object {
          "message": "Value must be a plain object",
        },
        "objectProperties": Object {
          "properties": Object {
            "address": Object {
              "notNull": Object {
                "message": "Value is required",
              },
              "notUndefined": Object {
                "message": "Value is required",
              },
              "objectPlain": Object {
                "message": "Value must be a plain object",
              },
              "objectProperties": Object {
                "properties": Object {
                  "line_1": Object {
                    "stringMaxLength": Object {
                      "length": 500,
                    },
                    "stringMinLength": Object {
                      "length": 10,
                    },
                  },
                },
              },
            },
            "name": Object {
              "notNull": Object {
                "message": "Value is required",
              },
              "notUndefined": Object {
                "message": "Value is required",
              },
            },
          },
        },
      },
      "title": Object {
        "notNull": Object {
          "message": "Value is required",
        },
        "notUndefined": Object {
          "message": "Value is required",
        },
      },
    },
  },
}
`;
